### PR TITLE
Implement LoadPairVector64 and LoadPairVector128

### LINKED ISSUE #39243: [Arm64] LoadPairVector64 and LoadPairVector128
```C#
namespace System.Runtime.Intrinsics.Arm
{
    public static class AdvSimd
    {
        public static unsafe (Vector64<byte> Value1,    Vector64<byte> Value2)    LoadPairVector64(byte*   address);
        public static unsafe (Vector64<sbyte> Value1,   Vector64<sbyte> Value2)   LoadPairVector64(sbyte*  address);
        public static unsafe (Vector64<short> Value1,   Vector64<short> Value2)   LoadPairVector64(short*  address);
        public static unsafe (Vector64<ushort> Value1,  Vector64<ushort> Value2)  LoadPairVector64(ushort* address);
        public static unsafe (Vector64<int> Value1,     Vector64<int> Value2)     LoadPairVector64(int*    address);
        public static unsafe (Vector64<uint> Value1,    Vector64<uint> Value2)    LoadPairVector64(uint*   address);
        public static unsafe (Vector64<float> Value1,   Vector64<float> Value2)   LoadPairVector64(float*  address);

        public static unsafe (Vector128<byte> Value1,   Vector128<byte> Value2)   LoadPairVector128(byte*   address);
        public static unsafe (Vector128<sbyte> Value1,  Vector128<sbyte> Value2)  LoadPairVector128(sbyte*  address);
        public static unsafe (Vector128<short> Value1,  Vector128<short> Value2)  LoadPairVector128(short*  address);
        public static unsafe (Vector128<ushort> Value1, Vector128<ushort> Value2) LoadPairVector128(ushort* address);
        public static unsafe (Vector128<int> Value1,    Vector128<int> Value2)    LoadPairVector128(int*    address);
        public static unsafe (Vector128<uint> Value1,   Vector128<uint> Value2)   LoadPairVector128(uint*   address);
        public static unsafe (Vector128<long> Value1,   Vector128<long> Value2)   LoadPairVector128(long*   address);
        public static unsafe (Vector128<ulong> Value1,  Vector128<ulong> Value2)  LoadPairVector128(ulong*  address);
        public static unsafe (Vector128<float> Value1,  Vector128<float> Value2)  LoadPairVector128(float*  address);

        public static unsafe (Vector64<int> Value1,     Vector64<int> Value2)     LoadPairScalarVector64(int*  address);
        public static unsafe (Vector64<uint> Value1,    Vector64<uint> Value2)    LoadPairScalarVector64(uint* address);
        public static unsafe (Vector64<long> Value1,    Vector64<long> Value2)    LoadPairScalarVector64(long*  address);
        public static unsafe (Vector64<ulong> Value1,   Vector64<ulong> Value2)   LoadPairScalarVector64(ulong* address);
        public static unsafe (Vector64<float> Value1,   Vector64<float> Value2)   LoadPairScalarVector64(float* address);
        public static unsafe (Vector64<double> Value1,  Vector64<double> Value2)  LoadPairScalarVector64(double* address);

        public static unsafe (Vector64<byte> Value1,    Vector64<byte> Value2)    LoadPairVector64NonTemporal(byte*   address);
        public static unsafe (Vector64<sbyte> Value1,   Vector64<sbyte> Value2)   LoadPairVector64NonTemporal(sbyte*  address);
        public static unsafe (Vector64<short> Value1,   Vector64<short> Value2)   LoadPairVector64NonTemporal(short*  address);
        public static unsafe (Vector64<ushort> Value1,  Vector64<ushort> Value2)  LoadPairVector64NonTemporal(ushort* address);
        public static unsafe (Vector64<int> Value1,     Vector64<int> Value2)     LoadPairVector64NonTemporal(int*    address);
        public static unsafe (Vector64<uint> Value1,    Vector64<uint> Value2)    LoadPairVector64NonTemporal(uint*   address);
        public static unsafe (Vector64<float> Value1,   Vector64<float> Value2)   LoadPairVector64NonTemporal(float*  address);

        public static unsafe (Vector128<byte> Value1,   Vector128<byte> Value2)   LoadPairVector128NonTemporal(byte*   address);
        public static unsafe (Vector128<sbyte> Value1,  Vector128<sbyte> Value2)  LoadPairVector128NonTemporal(sbyte*  address);
        public static unsafe (Vector128<short> Value1,  Vector128<short> Value2)  LoadPairVector128NonTemporal(short*  address);
        public static unsafe (Vector128<ushort> Value1, Vector128<ushort> Value2) LoadPairVector128NonTemporal(ushort* address);
        public static unsafe (Vector128<int> Value1,    Vector128<int> Value2)    LoadPairVector128NonTemporal(int*    address);
        public static unsafe (Vector128<uint> Value1,   Vector128<uint> Value2)   LoadPairVector128NonTemporal(uint*   address);
        public static unsafe (Vector128<long> Value1,   Vector128<long> Value2)   LoadPairVector128NonTemporal(long*   address);
        public static unsafe (Vector128<ulong> Value1,  Vector128<ulong> Value2)  LoadPairVector128NonTemporal(ulong*  address);
        public static unsafe (Vector128<float> Value1,  Vector128<float> Value2)  LoadPairVector128NonTemporal(float*  address);

        public static unsafe (Vector64<int> Value1,     Vector64<int> Value2)     LoadPairScalarVector64NonTemporal(int*  address);
        public static unsafe (Vector64<uint> Value1,    Vector64<uint> Value2)    LoadPairScalarVector64NonTemporal(uint* address);
        public static unsafe (Vector64<long> Value1,    Vector64<long> Value2)    LoadPairScalarVector64NonTemporal(long*  address);
        public static unsafe (Vector64<ulong> Value1,   Vector64<ulong> Value2)   LoadPairScalarVector64NonTemporal(ulong* address);
        public static unsafe (Vector64<float> Value1,   Vector64<float> Value2)   LoadPairScalarVector64NonTemporal(float* address);
        public static unsafe (Vector64<double> Value1,  Vector64<double> Value2)  LoadPairScalarVector64NonTemporal(double* address);
```

Extracted from #37014, see https://github.com/dotnet/runtime/issues/37014#issuecomment-657896555

### PR DESCRIPTION
**Pull request by echesakov**
Resolves #39243

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@dotnet/jit-contrib @tannergooding PTAL

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -706,14 +706,35 @@ int GenTree::GetRegisterDstCount(Compiler* compiler) const
...
+            case NI_AdvSimd_Arm64_LoadPairVector64NonTemporal:
+            case NI_AdvSimd_Arm64_LoadPairVector128:
+            case NI_AdvSimd_Arm64_LoadPairVector128NonTemporal:
+                return 2;
+
+            default:
+                unreached();
+        }
+#elif defined(TARGET_XARCH)
         return 2;
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.cpp:733]: nit: would be good to explicitly cover the intrinsic IDs for xarch as well (definitely could be a separate PR).

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -17689,6 +17710,12 @@ bool GenTree::isContainableHWIntrinsic() const
     {
         case NI_Vector64_get_Zero:
         case NI_Vector128_get_Zero:
+        case NI_AdvSimd_Arm64_LoadPairScalarVector64:
+        case NI_AdvSimd_Arm64_LoadPairScalarVector64NonTemporal:
+        case NI_AdvSimd_Arm64_LoadPairVector64:
+        case NI_AdvSimd_Arm64_LoadPairVector64NonTemporal:
+        case NI_AdvSimd_Arm64_LoadPairVector128:
+        case NI_AdvSimd_Arm64_LoadPairVector128NonTemporal:
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.cpp:17718]: I'll probably see later in the review, but under what conditions are these containable?

I didn't think Arm64 really had `ins reg, [mem]` operations outside atomic instructions; particularly for `non-temporal` operations...

File: src/coreclr/jit/gentree.h [hunk 3]
```diff
@@ -8039,12 +8039,12 @@ inline bool GenTree::IsMultiRegNode() const
         return true;
     }
 #endif // FEATURE_MULTIREG_RET
-#if defined(TARGET_XARCH) && defined(FEATURE_HW_INTRINSICS)
-    if (OperIs(GT_HWINTRINSIC))
+
+    if (OperIsHWIntrinsic())
     {
-        return (TypeGet() == TYP_STRUCT);
+        return TypeIs(TYP_STRUCT);
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.h:8045]: I wonder how well this will hold in the future? That is, are we expecting things that return `TYP_STRUCT `to always be multi-reg?

There are cases like say `System.Half` where we'll eventually need to add support and we'll need to treat it as `TYP_HALF` or recognize it some other way if we don't want it to be an issue.

I wonder if we should track this as a `flag` rather than by type just to be safe?

File: src/coreclr/jit/gentree.h [hunk 4]
```diff
@@ -8087,13 +8087,33 @@ inline unsigned GenTree::GetMultiRegCount()
...
+                return 2;
+
+            default:
+                unreached();
+        }
+#elif defined(TARGET_XARCH)
         return 2;
-    }
 #endif
+    }
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.h:8115]: This looks to be the same logic as in `gentree.cpp` above. Should it be factored out into a shared helper or are there conditions where they won't/shouldn't be in sync?

File: src/coreclr/jit/gentree.h [hunk 5]
```diff
@@ -8213,15 +8234,26 @@ inline var_types GenTree::GetRegTypeByIndex(int regIndex)
...
+        assert(TypeGet() == TYP_STRUCT);
+#ifdef TARGET_ARM64
+        if (AsHWIntrinsic()->GetSimdSize() == 16)
+        {
+            return TYP_SIMD16;
+        }
+        else
+        {
+            assert(AsHWIntrinsic()->GetSimdSize() == 8);
+            return TYP_SIMD8;
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.h:8248]: Do we have any cases of "arg size is 8" but "return size is 16" or vice-versa?

I know some instructions fit that bill, I'm not sure if any of the multi-reg cases will

File: src/coreclr/jit/gentree.h (hunk 3 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.h:8045]: Ah, looks like we have a flag. Maybe we should assert it or use it here instead?

File: src/coreclr/jit/gentree.cpp (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp:17718]: Actually, I don't see where the containment handling is happening for this. I don't see any changes to lowering

**Comment by tannergooding** [comment]
Changes generally LGTM. Left some comments/questions on a couple bits on how we expect certain checks to work long term.

Would be great to see a couple diffs/codegen examples particularly for the library changes.

File: src/coreclr/jit/gentree.cpp (hunk 1 above)
Comment (echesakovMSFT) [reply src/coreclr/jit/gentree.cpp:733]: Right, given that none are supported currently on X86 - I can replace this with `unreached()` and update with the `switch` when working on `MultiplyNoFlags2` (or whatever name we decided)

File: src/coreclr/jit/gentree.h (hunk 3 above)
Comment (echesakovMSFT) [reply src/coreclr/jit/gentree.h:8045]: Sure, I can use a flag here.

File: src/coreclr/jit/gentree.h (hunk 5 above)
Comment (echesakovMSFT) [reply src/coreclr/jit/gentree.h:8248]: I don't see an example on Arm64 when this wouldn't hold.
`ld[1-4]` should be similar to `ldp`.

As for `tbl` and `tbx`:
```
TBX <Vd>.<Ta>, { <Vn>.16B, <Vn+1>.16B, <Vn+2>.16B }, <Vm>.<Ta>
```
the return value is going to be single-reg but the first source operand is multi-reg and composed of `Vector128<byte>`.

**Comment by echesakovMSFT** [comment]
> Would be great to see a couple diffs/codegen examples particularly for the library changes.

@tannergooding There are suboptimalities in the code using `LoadPairVector` in the libraries. 

To fix that I would need to implement https://github.com/dotnet/runtime/pull/64863 and https://github.com/dotnet/runtime/pull/64857.

If I apply these changes on top of this PR the code diffs would look as expected:

**GetIndexOfFirstCharToEncodeAdvSimd64**
```diff
@@ -69,10 +73,8 @@ G_M52035_IG03:
                                                ;; bbWeight=0.50 PerfScore 0.25
 G_M52035_IG04:
             add     x6, x1, x4, LSL #1
-            ld1     {v20.8h}, [x6]
+            ldp     q20, q21, [x6]
             sqxtun  v20.8b, v20.8h
-            add     x6, x6, #16
-            ld1     {v21.8h}, [x6]
             sqxtun2 v20.16b, v21.8h
             and     v21.16b, v20.16b, v16.16b
             tbl     v21.16b, {v19.16b}, v21.16b
@@ -87,7 +89,7 @@ G_M52035_IG04:
             add     x4, x4, #16
             cmp     x4, x5
             blo     G_M52035_IG04
-                                               ;; bbWeight=4    PerfScore 100.00
+                                               ;; bbWeight=4    PerfScore 86.00
```

**GetIndexOfFirstNonAsciiByte_Intrinsified**
```diff
@@ -208,9 +211,8 @@ G_M18966_IG11:
@@ -208,9 +211,7 @@ G_M18966_IG11:
             sub     x22, x0, #32
                                                ;; bbWeight=0.50 PerfScore 1.75
 G_M18966_IG12:
-            ld1     {v16.16b}, [x19]
-            add     x0, x19, #16
-            ld1     {v10.16b}, [x0]
+            ldp     q16, q9, [x19]
             sshr    v16.16b, v16.16b, #7
             and     v16.16b, v16.16b, v8.16b
             addp    v16.16b, v16.16b, v16.16b
```

Here is my plan:
1) Undo the changes to these methods
2) Merge this PR as is
3) Finish the above-mentioned two PRs
4) Follow-up with the libraries changes

I will keep the changes to **BitArray:CopyTo** though
```diff
@@ -413,18 +414,16 @@ G_M40488_IG18:
             zip1    v19.16b, v18.16b, v18.16b
             and     v19.16b, v19.16b, v17.16b
             umin    v19.16b, v19.16b, v16.16b
-            st1     {v19.16b}, [x3]
             zip2    v18.16b, v18.16b, v18.16b
             and     v18.16b, v18.16b, v17.16b
             umin    v18.16b, v18.16b, v16.16b
-            add     x3, x3, #16
-            st1     {v18.16b}, [x3]
+            stp     q19, q18, [x3]
             add     w1, w1, #32
             add     w3, w1, #32
             ldr     w4, [x19,#16]
             cmp     w3, w4
             bls     G_M40488_IG18
```

**Comment by echesakovMSFT** [comment]
Can someone on Mono team to sign off on the relevant changes, please? 

For context: 15e56a002f5c740030796ddb6dd12874e6331d84 was implemented by @imhameed during my initial attempt to get these changes in #52424

**Comment by echesakovMSFT** [comment]
@dotnet/jit-contrib PTAL

**Review by BruceForstall — APPROVED** [review]
LGTM

**Review by tannergooding — APPROVED** [review]

**Comment by imhameed** [comment]
> Can someone on Mono team to sign off on the relevant changes, please?
> 
> For context: [15e56a0](https://github.com/dotnet/runtime/commit/15e56a002f5c740030796ddb6dd12874e6331d84) was implemented by @imhameed during my initial attempt to get these changes in #52424

I'm 23 hours late to this but: the changes still look good to me (although I don't know if me signing off on my own implementation is kosher)
